src/main/java/org/traccar/protocol/ZrProtocolDecoder.java

```diff
@@ -0,0 +1,323 @@
+            int len = buf.readUnsignedShort();
+            ByteBuf valueBuf = buf.readSlice(len);
+
+            switch (tag) {
+                case 0x2095:
+                    position.setAccuracy((double) valueBuf.readUnsignedShort() / 100);
+                    position.setSpeed(valueBuf.readUnsignedShort());
+                    position.setCourse(valueBuf.readUnsignedShort());
+                    position.setAltitude(valueBuf.readInt());
+                    position.setLongitude(valueBuf.readInt() * 0.000001);
+                    position.setLatitude(valueBuf.readInt() * 0.000001);
+                    long l = valueBuf.readUnsignedInt(); // UTC second time
+                    Calendar calendar = Calendar.getInstance((TimeZone) deviceSession.get(DeviceSession.KEY_TIMEZONE));
+                    calendar.setTimeInMillis(l * 1000L);
+                    position.setTime(calendar.getTime());
+                    break;
+                case 0x240c:
+                    position.set(Position.KEY_ODOMETER, valueBuf.readUnsignedInt());
+                    break;
+                case 0x2310:
+                    long value = valueBuf.readLong();
+                    position.set(Position.KEY_ALARM, decode2310(value));
+
+                    if (BitUtil.check(value, 12)) {
+                        deviceSession.set(STATE_ACC, ACC_ON);
+                    }
+                    if (BitUtil.check(value, 13)) {
+                        deviceSession.set(STATE_ACC, ACC_OFF);
+                    }
+                    break;
+            }
+        }
+
+        if (position.getLatitude() != 0 && position.getLongitude() != 0) {
+            position.setValid(true);
+        } else {
+            return null;
+        }
+
+        Integer accState = deviceSession.get(STATE_ACC);
+        position.set(Position.KEY_IGNITION, accState != null && accState.equals(ACC_ON));
+        return position;
+    }
+
+    private String decode2310(long value) {
+        if (BitUtil.check(value, 3)) {
+            return Position.ALARM_POWER_OFF;
+        }
+        if (BitUtil.check(value, 4)) {
+            return Position.ALARM_LOW_POWER;
+        }
+        if (BitUtil.check(value, 22)) {
+            return Position.ALARM_OVERSPEED;
+        }
+        if (BitUtil.check(value, 41)) {
+            return Position.ALARM_TOW;
+        }
+        if (BitUtil.check(value, 42)) {
+            return Position.ALARM_FALL_DOWN;
+        }
+        if (BitUtil.check(value, 43)) {
+            return Position.ALARM_ACCIDENT;
+        }
+        if (BitUtil.check(value, 44)) {
+            return Position.ALARM_VIBRATION;
+        }
+        if (BitUtil.check(value, 50)) {
+            return Position.ALARM_REMOVING;
+        }
+
+        return null;
+    }
+
+    private String decodeId(ByteBuf id) {
+        String str = ByteBufUtil.hexDump(id);
+        int i = 0;
+        while (i < str.length() && str.charAt(i) == '0') {
+            i++;
+        }
```

Comment:
Why are we dropping zeroes?

Reply:
Because in practice, the device number is only 15 digits fixed, but the protocol output has 20 digits. We want to remove the leading zeros to make the device number look clearer.